nio/client/async_client.py

```diff
@@ -522,39 +563,29 @@ def add_response_callback(
         """
         while True:
             try:
-                responses = []
-
-                responses.append(
-                    await self.sync(timeout, sync_filter, since, full_state)
-                )
-
-                full_state = None
-                since = None
-
-                responses += await self.send_to_device_messages()
+                coros = [
+                    self.sync(timeout, sync_filter, since, full_state),
+                    self.send_to_device_messages()
+                ]
 
                 if self.should_upload_keys:
-                    responses.append(await self.keys_upload())
+                    coros.append(self.keys_upload())
 
                 if self.should_query_keys:
-                    responses.append(await self.keys_query())
+                    coros.append(self.keys_query())
 
-                await self.run_response_callbacks(responses)
+                for response in asyncio.as_completed(coros):
```

Comment:
If we create a task for `sync_forever` and cancel it later on the asyncio run-time complains that some task here gets destroyed but it was pending, that is the `coros` here don't get canceled.

One way handle this is to turn the coroutines into tasks using `ensure_future()` and cancel them when we catch the `CancelledError`